File: 0238-product-of-array-except-self/238.cpp
class Solution {
  public:
    vector<int> productExceptSelf(vector<int>& nums) {
        /*
         * Approach:
         * 1. Brute Force                     -- T.C: O(N^2), S.C: O(1)
         *      - For each element, multiply all others.
         * 2. Prefix + Suffix Arrays          -- T.C: O(N),   S.C: O(N)
         *      - Build prefix and suffix arrays, multiply at each index.
         * 3. Prefix + Suffix Space Optimized -- T.C: O(N),   S.C: O(1)
         *      - Use output array for prefix pass, running suffix for second pass.
         */

        // Approach 3 -- Prefix + Suffix Space Optimized
        int         n = nums.size();
        vector<int> ans(n, 1);

        // Pass 1: fill ans with prefix products
        for (int i = 1; i < n; i++) ans[i] = ans[i - 1] * nums[i - 1];

        // Pass 2: multiply running suffix into ans
        int suffix = 1;
        for (int i = n - 1; i >= 0; i--) {
            ans[i] *= suffix;
            suffix *= nums[i];
        }
        return ans;

        /* Approach 2 -- Prefix + Suffix Arrays
        int         n = nums.size();
        vector<int> prefix(n, 1), suffix(n, 1), ans(n);
        for (int i = 1; i < n; i++)
            prefix[i] = prefix[i - 1] * nums[i - 1];
        for (int i = n - 2; i >= 0; i--)
            suffix[i] = suffix[i + 1] * nums[i + 1];
        for (int i = 0; i < n; i++)
            ans[i] = prefix[i] * suffix[i];
        return ans;
        */

        /* Approach 1 -- Brute Force
        int         n = nums.size();
        vector<int> ans(n, 1);
        for (int i = 0; i < n; i++) {
            int product = 1;
            for (int j = 0; j < n; j++)
                if (i != j)
                    product *= nums[j];
            ans[i] = product;
        }
        return ans;
        */
    }
};

File: 0238-product-of-array-except-self/238.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        """
        Approach:
        1. Brute Force                     -- T.C: O(N^2), S.C: O(1)
             - For each element, multiply all others.
        2. Prefix + Suffix Arrays          -- T.C: O(N),   S.C: O(N)
             - Build prefix and suffix arrays, multiply at each index.
        3. Prefix + Suffix Space Optimized -- T.C: O(N),   S.C: O(1)
             - Use output array for prefix pass, running suffix for second pass.
        """

        # Approach 3 -- Prefix + Suffix Space Optimized
        n = len(nums)
        ans = [1] * n

        # Pass 1: fill ans with prefix products
        for i in range(1, n):
            ans[i] = ans[i - 1] * nums[i - 1]

        # Pass 2: multiply running suffix into ans
        suffix = 1
        for i in range(n - 1, -1, -1):
            ans[i] *= suffix
            suffix *= nums[i]
        return ans

        # Approach 2 -- Prefix + Suffix Arrays
        # n      = len(nums)
        # prefix = [1] * n
        # suffix = [1] * n
        # for i in range(1, n):
        #     prefix[i] = prefix[i - 1] * nums[i - 1]
        # for i in range(n - 2, -1, -1):
        #     suffix[i] = suffix[i + 1] * nums[i + 1]
        # return [prefix[i] * suffix[i] for i in range(n)]

        # Approach 1 -- Brute Force
        # n   = len(nums)
        # ans = [1] * n
        # for i in range(n):
        #     product = 1
        #     for j in range(n):
        #         if i != j:
        #             product *= nums[j]
        #     ans[i] = product
        # return ans
